### ISSUE TITLE
Order-dependent segfault in `set_add`

### ISSUE THREAD

**Issue by mkeeter** [prove-rs/z3.rs#7]
The following code works fine:
```
   let cfg = Config::new();
    let ctx = Context::new(&cfg);
    let int_sort = ctx.int_sort();
    let int_set_sort = ctx.set_sort(&int_sort);
    let set = ctx.named_const("set", &int_set_sort);
    let a = ctx.named_int_const("a");
    let result = set.set_add(&a);
```

However, if you swap the `let set...` and `let a...` lines (so that `a` is defined first), it crashes with `SIGSEGV`.  Oddly, running in `lldb` doesn't crash: it prints `Error: type error` and exits with return code 1.

I'm running `z3 0.3.1` and `z3-sys 0.4.0`.

**Comment by waywardmonkeys** [comment]
Thanks for reporting this. This is interesting as it seems like it would be an error in refcounting. The question is: Is it here in the Rust binding or in the Z3 API implementation itself?

What I find interesting is that the Z3 API implementation doesn't appear to be doing much here in terms of refcounting, while inserting into an AST map does.

It might be useful to convert this over to something using the C API and bring it up upstream. I'll try to do that within the next day or two, but I'm massively jetlagged, so if I forget, please feel free to ping me again.

**Comment by mkeeter** [comment]
Here's equivalent code in C, as far as I can tell.  This is the crashing-in-Rust configuration (defining `a` before `set`):

```
#include <stdio.h>
#include <z3.h>

int main(int argc, char** argv) {
    Z3_config cfg = Z3_mk_config();
    Z3_context ctx = Z3_mk_context(cfg);

    Z3_sort int_sort = Z3_mk_int_sort(ctx);
    Z3_sort int_set_sort = Z3_mk_set_sort(ctx, int_sort);

    Z3_symbol a_name = Z3_mk_string_symbol(ctx, "a");
    Z3_ast a = Z3_mk_const(ctx, a_name, int_sort);

    Z3_symbol set_name = Z3_mk_string_symbol(ctx, "set");
    Z3_ast set = Z3_mk_const(ctx, set_name, int_set_sort);

    Z3_ast result = Z3_mk_set_add(ctx, set, a);
    printf("%s", Z3_ast_to_string(ctx, result));
}
```

This runs fine using the C API, which suggests that the issue is in the Rust bindings themselves.  I also did the same thing in Python without a crash.

**Comment by waywardmonkeys** [comment]
Almost ... but there's a major difference. The Rust bindings are using `Z3_mk_context_rc` instead, which requires that some extra refcounting work be done. Unfortunately, I'm still seeing a crash in invoking `Z3_mk_set_add` without doing any `Z3_dec_ref` calls, but when I swap the `a` and `set` construction, I see it succeed.

This is pretty interesting!

I'll submit this test case upstream and see what they say:

```c
#include <stdio.h>
#include <z3.h>

int main(int argc, char** argv) {
    Z3_config cfg = Z3_mk_config();
    Z3_context ctx = Z3_mk_context_rc(cfg);

    Z3_sort int_sort = Z3_mk_int_sort(ctx);
    Z3_sort int_set_sort = Z3_mk_set_sort(ctx, int_sort);

    Z3_symbol set_name = Z3_mk_string_symbol(ctx, "set");
    Z3_ast set = Z3_mk_const(ctx, set_name, int_set_sort);
    Z3_inc_ref(ctx, set);

    Z3_symbol a_name = Z3_mk_string_symbol(ctx, "a");
    Z3_ast a = Z3_mk_const(ctx, a_name, int_sort);
    Z3_inc_ref(ctx, a);

    Z3_ast result = Z3_mk_set_add(ctx, set, a);
    Z3_inc_ref(ctx, result);
    printf("%s", Z3_ast_to_string(ctx, result));
}
```

**Comment by waywardmonkeys** [comment]
The issue is that we're not doing a `Z3_inc_ref` on the `Z3_sort` objects that we create. That's wrong on the part of the Rust bindings.

I have to sleep now (3:15am), but I'll look at this in the morning after I wake up.

**Comment by mkeeter** [comment]
Thanks for the quick follow-up!

For now, I've added a load-bearing dummy variable – constructing any `set` before the first `int` seems to work around the crash.

**Comment by waywardmonkeys** [comment]
This should be fixed in the version 0.3.2 that I just published. Please re-open if it isn't!
